File: backend/src/main/java/com/uet/VolunteerHub/service/RequestService.java
package com.uet.VolunteerHub.service;

import com.uet.VolunteerHub.dto.Request.RequestCreateDTO;
import com.uet.VolunteerHub.dto.Request.RequestReadDTO;
import com.uet.VolunteerHub.dto.Request.RequestReviewDTO;
import com.uet.VolunteerHub.entity.Account;
import com.uet.VolunteerHub.entity.Request;
import com.uet.VolunteerHub.enums.RequestStatus;
import com.uet.VolunteerHub.enums.UserRole;
import com.uet.VolunteerHub.exception.ResourceAlreadyExistsException;
import com.uet.VolunteerHub.exception.ResourceNotFoundException;
import com.uet.VolunteerHub.repository.AccountRepository;
import com.uet.VolunteerHub.repository.RequestRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
public class RequestService {

    private final RequestRepository requestRepository;
    private final AccountRepository accountRepository;

    @Autowired
    public RequestService(RequestRepository requestRepository, AccountRepository accountRepository) {
        this.requestRepository = requestRepository;
        this.accountRepository = accountRepository;
    }

    private RequestReadDTO mapToRequestReadDTO(Request request) {
        return RequestReadDTO.builder()
                .requestId(request.getRequestId())
                .accountId(request.getAccount().getAccountId())
                .username(request.getAccount().getUsername())
                .email(request.getAccount().getEmail())
                .status(request.getStatus())
                .reason(request.getReason())
                .adminResponse(request.getAdminResponse())
                .createdAt(request.getCreatedAt())
                .updatedAt(request.getUpdatedAt())
                .build();
    }

    @Transactional
    public RequestReadDTO createRequest(Account account, RequestCreateDTO requestCreateDTO) {
        if (account.getRole() == UserRole.MANAGER || account.getRole() == UserRole.ADMIN) {
            throw new IllegalArgumentException("You already have manager or admin privileges");
        }
        if (requestRepository.existsByAccountAndStatus(account, RequestStatus.WAITING)) {
            throw new ResourceAlreadyExistsException("You already have a pending request");
        }

        Request request = Request.builder()
                .account(account)
                .status(RequestStatus.WAITING)
                .reason(requestCreateDTO.getReason())
                .build();

        request = requestRepository.save(request);
        return mapToRequestReadDTO(request);
    }

    @Transactional
    public RequestReadDTO reviewRequest(Long requestId, Account admin, RequestReviewDTO requestReviewDTO) {
        Request request = requestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Request with ID " + requestId + " not found"));
        if (request.getStatus() != RequestStatus.WAITING) {
            throw new IllegalArgumentException("This request has already been reviewed");
        }
        if (requestReviewDTO.getStatus() == RequestStatus.WAITING) {
            throw new IllegalArgumentException("Cannot set status to WAITING");
        }
        request.setStatus(requestReviewDTO.getStatus());
        request.setAdminResponse(requestReviewDTO.getAdminResponse());
        if (requestReviewDTO.getStatus() == RequestStatus.APPROVED) {
            Account userAccount = request.getAccount();
            userAccount.setRole(UserRole.MANAGER);
            accountRepository.save(userAccount);
        }
        request = requestRepository.save(request);
        return mapToRequestReadDTO(request);
    }

    @Transactional
    public void cancelRequest(Long requestId, Account account) {
        Request request = requestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Request with ID " + requestId + " not found"));
        if (!request.getAccount().getAccountId().equals(account.getAccountId())) {
            throw new IllegalArgumentException("You can only cancel your own requests");
        }
        if (request.getStatus() != RequestStatus.WAITING) {
            throw new IllegalArgumentException("You can only cancel pending requests");
        }
        requestRepository.delete(request);
    }

    @Transactional
    public Page<RequestReadDTO> getAllRequests(Pageable pageable) {
        Page<Request> requests = requestRepository.findAll(pageable);
        return requests.map(this::mapToRequestReadDTO);
    }

    @Transactional
    public Page<RequestReadDTO> getRequestsByStatus(RequestStatus status, Pageable pageable) {
        Page<Request> requests = requestRepository.findByStatus(status, pageable);
        return requests.map(this::mapToRequestReadDTO);
    }

    @Transactional
    public Page<RequestReadDTO> getMyRequests(Account account, Pageable pageable) {
        Page<Request> requests = requestRepository.findByAccount(account, pageable);
        return requests.map(this::mapToRequestReadDTO);
    }

    @Transactional
    public RequestReadDTO getRequestById(Long requestId, Account account) {
        Request request = requestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Request with ID " + requestId + " not found"));
        
        // Kiểm tra quyền sở hữu: chỉ người tạo request mới được xem
        if (!request.getAccount().getAccountId().equals(account.getAccountId())) {
            throw new IllegalArgumentException("You can only view your own requests");
        }
        
        return mapToRequestReadDTO(request);
    }

    @Transactional
    public RequestReadDTO getRequestByIdForAdmin(Long requestId) {
        Request request = requestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Request with ID " + requestId + " not found"));
        return mapToRequestReadDTO(request);
    }

    @Transactional
    public long countPendingRequests() {
        return requestRepository.countByStatus(RequestStatus.WAITING);
    }
}

File: frontend/src/api/services/request.service.js
import api from '../api';

export const requestService = {
  // ========== USER ENDPOINTS ==========
  
  // POST /api/requests - Create a new manager request
  async createRequest(reason) {
    const response = await api.post('/requests', { reason });
    return response.data;
  },

  // GET /api/requests/me - Get current user's requests
  async getMyRequests(page = 0, size = 10) {
    const response = await api.get('/requests/me', {
      params: { page, size }
    });
    return response.data;
  },

  // GET /api/requests/{requestId} - Get request by ID
  async getRequestById(requestId) {
    const response = await api.get(`/requests/${requestId}`);
    return response.data;
  },

  // DELETE /api/requests/{requestId} - Cancel a pending request
  async cancelRequest(requestId) {
    await api.delete(`/requests/${requestId}`);
  },

  // ========== ADMIN ENDPOINTS ==========

  // GET /api/admin/requests - Get all requests
  async getAllRequests(page = 0, size = 10) {
    const response = await api.get('/admin/requests', {
      params: { page, size }
    });
    return response.data;
  },

  // GET /api/admin/requests/status/{status} - Get requests by status
  async getRequestsByStatus(status, page = 0, size = 10) {
    const response = await api.get(`/admin/requests/status/${status}`, {
      params: { page, size }
    });
    return response.data;
  },

  // GET /api/admin/requests/pending-count - Get count of pending requests
  async getPendingCount() {
    const response = await api.get('/admin/requests/pending-count');
    return response.data;
  },

  // PATCH /api/admin/requests/{requestId}/review - Approve or reject request
  async reviewRequest(requestId, status, adminResponse = '') {
    const response = await api.patch(`/admin/requests/${requestId}/review`, {
      status,
      adminResponse
    });
    return response.data;
  },

  // GET /api/admin/requests/{requestId} - Get request details (admin)
  async getRequestByIdAdmin(requestId) {
    const response = await api.get(`/admin/requests/${requestId}`);
    return response.data;
  }
};
